### PR TITLE
Add optional PKCE support and local validation docs (PROJQUAY-9281)

### PR DESCRIPTION
**Pull request by HammerMeetNail**
- Implement optional PKCE for OIDC login flows:
  - Add oauth/pkce.py with code_verifier and S256 code_challenge helpers
  - Extend OAuthService.get_auth_url and exchange_code to accept extra auth/token params and allow omitting client_secret for public clients
  - Update OIDCLoginService to plumb PKCE (USE_PKCE, PKCE_METHOD, PUBLIC_CLIENT) and pass code_verifier during token exchange
  - Wire endpoints to generate/store code_verifier in session, include code_challenge(+_method) on authorization, and use code_verifier on callback/token exchange
- Add plans/oidc_pkcd.md with phased implementation plan
- Add plans/oidc_pkce_validation.md with QA step-by-step local validation (Keycloak + Quay)
- Local dev config: Keycloak provider example and USE_PKCE: true

Notes:
- OIDC_SERVER requires trailing slash; use host.containers.internal with podman
- PKCE disabled by default; enable per provider via USE_PKCE

### CODE REVIEW

File: oauth/pkce.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        raise ValueError("PKCE code_verifier length must be between 43 and 128 characters")
+    return "".join(secrets.choice(_UNRESERVED) for _ in range(length))
+
+
+def code_challenge(verifier: str, method: str = "S256") -> str:
+    if method.upper() == "PLAIN":
+        return verifier
+    if method.upper() != "S256":
+        raise ValueError("Unsupported PKCE method: %s" % method)
+    digest = hashlib.sha256(verifier.encode("ascii")).digest()
```
[the remark this answers is not in the document]
Comment (HammerMeetNail) [reply oauth/pkce.py:20]: This is not password hashing. The code is implementing RFC 7636 (PKCE for OAuth2), which specifically mandates using SHA256 for the S256 code challenge method. See https://datatracker.ietf.org/doc/html/rfc7636#section-4.2

**Comment by HammerMeetNail** [comment]
Review from gpt-5:

### Code Review: rfe-7998-oidc-pkce

#### Summary
- Adds optional PKCE support (S256 and plain) for OIDC login, with public-client support and session-based verifier handling.
- Extends `oauth/base.py` to accept extra auth/token params and omit client secret for public clients.
- Implements PKCE helpers in `oauth/pkce.py` and wires flows in `oauth/oidc.py`, `endpoints/api/user.py`, and `endpoints/oauth/login.py`.
- Adds extensive tests covering PKCE utilities, OIDC token/login flows, session verifier persistence, and base OAuth behavior.

#### Files reviewed
- `oauth/pkce.py`
- `oauth/oidc.py`
- `oauth/base.py`
- `endpoints/api/user.py` (ExternalLoginInformation)
- `endpoints/oauth/login.py`
- Tests: `oauth/test/test_*.py`, `endpoints/api/test/test_user_pkce.py`, `endpoints/oauth/test/test_login_pkce.py`
- Local dev config: `local-dev/stack/config.yaml`

### Test coverage
The added tests are strong and exercise key scenarios:
- PKCE utility correctness: verifier generation constraints and S256 challenge computation.
- OIDC token exchange with PKCE (verifier included) and without (failure when server requires it).
- Public client behavior: ensure client_secret is omitted during token request.
- Support for both S256 and plain methods.
- Session storage/retrieval of PKCE verifier across login, attach, and CLI flows.
- Base OAuth API changes: extra token params, omit client secret, form encoding.

Small gaps (optional):
- End-to-end route-level tests for `endpoints/oauth/login.py` callbacks (current tests simulate logic, not the registered blueprint routes).
- TTL/expiry handling for PKCE verifier is mentioned in plans but not implemented or tested.

### Potential bugs and edge cases
- Defensive session handling in callbacks:
  - In `endpoints/oauth/login.py` the code accesses `data.get("verifier")` without verifying `data` is a mapping. If the session value is corrupted (string/None), this could raise. Consider guarding with `hasattr(data, "get")` before access.

  Example area:
  ```
  data = session.pop(session_key, None)
  if data and data.get("verifier"):
      kwargs["code_verifier"] = data.get("verifier")
  ```

- Retry backoff duration in token exchange:
  - In `oauth/base.py`, `timeout = 5 / 1000` (5ms) between retries is too short to be effective for network blips. Recommend >= 0.5–1.0s or exponential backoff.

- Explicit Content-Type for form-encoded requests:
  - When `form_encode=True` in `oauth/base.py`, set `Content-Type: application/x-www-form-urlencoded` explicitly to avoid servers rejecting the request.

- Non-verified decode path uses key unnecessarily:
  - In `oauth/oidc.py` retry path, the non-verified decode (`verify_signature=False`) still fetches a key. Consider passing `key=None` there to avoid confusion.

- PKCE method validation:
  - `pkce_method()` returns arbitrary config; if misconfigured, authorize requests may advertise an unsupported method and token requests might fail in obscure ways. Consider validating to {"S256", "plain"} and failing fast on misconfig.

- Verifier TTL/reuse prevention:
  - Session structure includes a timestamp, but callbacks don’t enforce a maximum age or single-use beyond a `pop` (which is good for single-use). If TTL is desired, enforce in callbacks.

### Suggestions / improvements
1) `endpoints/oauth/login.py`:
   - Add a safe-guard when pulling the verifier from session:
     ```python
     data = session.pop(session_key, None)
     if data and hasattr(data, "get") and data.get("verifier"):
         kwargs["code_verifier"] = data.get("verifier")
     ```
   - Optional: Enforce TTL, e.g. ignore if `int(time.time()) - data.get("ts", 0) > 600`.

2) `oauth/base.py`:
   - Increase retry backoff and/or use exponential backoff (e.g., 0.5s, 1.0s, 2.0s).
   - When `form_encode=True`, set `headers["Content-Type"] = "application/x-www-form-urlencoded"`.

3) `oauth/oidc.py`:
   - In non-verified decode logging path, pass `key=None` to `decode(..., verify_signature=False)` to reflect the intent.
   - Validate `pkce_method()` against allowed set and raise a clear configuration error if invalid.

4) Tests (nice-to-have):
   - Add route-level tests for the registered OAuth blueprint callbacks (login/attach/cli) to cover pop-once session behavior end-to-end.
   - Add a negative test for invalid `PKCE_METHOD` configuration resulting in early failure.
   - Add tests for login/token error messages on missing `access_token` or `id_token` to guard against regressions.

### Local-dev config notes
- `local-dev/stack/config.yaml` enables PKCE and PUBLIC_CLIENT for the example OIDC provider. Ensure the IdP client is actually configured as public or confidential to match this setting; otherwise token exchange will fail.

### Overall
The PKCE feature is implemented cleanly with minimal, backward-compatible changes and strong test coverage. Addressing the small robustness items (session guard, realistic retry delay, explicit Content-Type, optional TTL, and PKCE method validation) would harden the implementation further.

**Comment by HammerMeetNail** [comment]
1. PR #4256 - pkce-foundation: https://github.com/quay/quay/pull/4256
    - Core PKCE implementation with session-verifier flow
  2. PR #4257 - pkce-tests: https://github.com/quay/quay/pull/4257
    - Comprehensive test suite covering all PKCE functionality
  3. PR #4258 - pkce-docs: https://github.com/quay/quay/pull/4258
    - Implementation and validation documentation

File: endpoints/oauth/login.py [hunk 2]
```diff
@@ -117,9 +117,15 @@ def callback_func():
 
         # Exchange the OAuth code for login information.
         code = request.values.get("code")
+        kwargs = {}
+        if hasattr(login_service, "pkce_enabled") and login_service.pkce_enabled():
+            session_key = f"_oauth_pkce_{login_service.service_id()}"
+            data = session.pop(session_key, None)
+            if data and hasattr(data, "get") and data.get("verifier"):
+                kwargs["code_verifier"] = data.get("verifier")
```
Comment (sridipta) [inline endpoints/oauth/login.py:125]: Same comment as I provided in the other PR (https://github.com/quay/quay/pull/4256).

PS. I see you have changes in the same lines in multiple branches that PRs open for. Careful with merge conflicts.

File: oauth/test/test_oidc.py [hunk 3]
```diff
@@ -399,6 +538,117 @@ def test_exchange_code_missingkey(
...
+
+
+def test_pkce_enabled_false_default(oidc_service):
+    # Verify False default when USE_PKCE not set
+    assert oidc_service.pkce_enabled() is False
+
+
+def test_pkce_method_s256_default(pkce_oidc_service):
+    # Verify S256 default method
+    assert pkce_oidc_service.pkce_method() == "S256"
```
Comment (sridipta) [inline oauth/test/test_oidc.py:553]: normalize the output of pkce_method() call (to say lowercase) to ensure case insensitivity. as we are directly reading the config file, there might be potential inconsitency in the casing of the strings "Plain" and "S256".

This needs to be done in multiple places.
